codec-http/src/main/java/io/netty/handler/codec/http/HttpServerUpgradeHandler.java

```diff
@@ -224,13 +224,24 @@ public HttpServerUpgradeHandler(SourceCodec sourceCodec, UpgradeCodecFactory upg
     @Override
     protected void decode(ChannelHandlerContext ctx, HttpObject msg, List<Object> out)
             throws Exception {
-        // Determine if we're already handling an upgrade request or just starting a new one.
-        handlingUpgrade |= isUpgradeRequest(msg);
+
         if (!handlingUpgrade) {
-            // Not handling an upgrade request, just pass it to the next handler.
-            ReferenceCountUtil.retain(msg);
-            out.add(msg);
-            return;
+            // Not handling an upgrade request yet. Check if we received a new upgrade request.
+            if (msg instanceof HttpRequest) {
+                HttpRequest req = (HttpRequest) msg;
+                if (req.headers().contains(HttpHeaderNames.UPGRADE) &&
+                    isUpgradeRequest(req)) {
```

Comment:
@trustin should we just always call `isUpgradeRequest` and move the `req.headers().contains(HttpHeaderNames.UPGRADE)` into `isUpgradeRequest` ? Imho that would make more sense

Reply:
I thought about that, but realized that it will not work if we returned `true` for the request without an `Upgrade` header anyway.

Reply:
(.. because this handler accesses the `Upgrade` header everywhere)

Reply:
maybe `shouldHandleUpgradeRequest(...)` ?

Reply:
Renamed. Thanks for inspiration.